kart/annotations/__init__.py

```diff
@@ -0,0 +1,82 @@
+import json
+import logging
+
+from sqlalchemy.exc import OperationalError
+
+from .db import annotations_session, KartAnnotation
+
+L = logging.getLogger(__name__)
+
+
+class DiffAnnotations:
+    def __init__(self, repo):
+        self.repo = repo
+
+    def _object_id(self, base_rs, target_rs):
+        # this is actually symmetric, so we can marginally increase hit rate by sorting first
+        tree_ids = sorted(rs.tree.id for rs in (base_rs, target_rs))
+        return f"{tree_ids[0]}...{tree_ids[1]}"
+
+    def store(self, *, base_rs, target_rs, annotation_type, data):
+        """
+        Stores a diff annotation to the repo's sqlite database,
+        and returns the annotation itself.
+
+        base_rs: base RepoStructure object for this diff (revA in a 'revA...revB' diff)
+        target_rs: target RepoStructure object for this diff (revB in a 'revA...revB' diff)
+        """
+        assert isinstance(data, dict)
+        object_id = self._object_id(base_rs, target_rs)
+        data = json.dumps(data)
+        try:
+            with annotations_session(self.repo) as session:
+                L.debug(
+                    "storing: %s for %s: %s",
+                    annotation_type,
+                    object_id,
+                    data,
+                )
+                session.add(
+                    KartAnnotation(
+                        object_id=object_id,
+                        annotation_type=annotation_type,
+                        data=data,
+                    )
+                )
+        except OperationalError as e:
+            # ignore errors from readonly databases.
+            if "readonly database" in str(e):
```

Comment:
if the directory isn't writeable does this raise "readonly database" or an IO error? Via the sqlite CLI I get `Error: unable to open database "doo.db": unable to open database file`

Reply:
have dealt with that now - it will switch to an in-memory database if the real db can't be created

Reply:
How's an in-memory database useful?

Reply:
> `# but we still need _some_ sqlalchemy session to yield, otherwise all the annotations code will have to handle this specifically. so we create a in-memory database and make it look readonly [so it isn't used]`

that makes a bit more sense.

All the low-level sqlite error handling seems to be the cache caller's responsibility which I'm not keen on, but at least it's only used in two places at the moment — can be improved later.